### PR TITLE
[Event Hubs] Buffered Producer Client Infrastructure (initial)

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The focus of these changes is to start building out the basic infrastructure of the buffered producer client and the tests that accompany it.  In this set of changes are:

- Structural tweaks
- Initial set of tests for validating constructors, event handlers, and close
- Fleshing out the dispose/close methods
- Extending with some internal aids for testing

# Related and References

- [EventHubBufferedProducerClient: Infrastructure and Functionality (#23480)](https://github.com/Azure/azure-sdk-for-net/issues/23480)

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs [hunk 1]
```diff
@@ -317,9 +340,8 @@ public EventHubBufferedProducerClient(string connectionString) : this(connection
         public EventHubBufferedProducerClient(string fullyQualifiedNamespace,
                                               string eventHubName,
                                               AzureNamedKeyCredential credential,
-                                              EventHubBufferedProducerClientOptions clientOptions = default) : this(clientOptions)
+                                              EventHubBufferedProducerClientOptions clientOptions = default) : this(fullyQualifiedNamespace, eventHubName, (object)credential, clientOptions)
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:343]: Is this cast required? 

```suggestion
                                              EventHubBufferedProducerClientOptions clientOptions = default) : this(fullyQualifiedNamespace, eventHubName, credential, clientOptions)
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs [hunk 2]
```diff
@@ -646,11 +646,64 @@ internal virtual Task ClearAsync(CancellationToken cancellationToken = default)
         ///   This method will automatically unregister the <see cref="SendEventBatchSucceededAsync"/> and <see cref="SendEventBatchFailedAsync"/> handlers.
         /// </remarks>
         ///
-        public virtual Task CloseAsync(bool flush = true,
-                                       CancellationToken cancellationToken = default)
+        public virtual async Task CloseAsync(bool flush = true,
+                                             CancellationToken cancellationToken = default)
         {
+            if (_isClosed)
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:652]: Is the `else` path ok to run twice in the event of a thread race?

**Review by christothes — APPROVED** [review]

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs (hunk 1 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:343]: Sadly, it was.   It kept giving me a message about trying to delegate to the same constructor.  I'm guessing the overload resolution was picking up on the stronger type match.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:343]: We had to do the same for `EventProcessor<T>`:  https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor.cs#L325

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs (hunk 2 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:652]: damn.  Good catch, it is not.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs [hunk 3]
```diff
@@ -646,11 +650,87 @@ internal virtual Task ClearAsync(CancellationToken cancellationToken = default)
...
-        public virtual Task CloseAsync(bool flush = true,
-                                       CancellationToken cancellationToken = default)
+        public virtual async Task CloseAsync(bool flush = true,
+                                             CancellationToken cancellationToken = default)
         {
-            _isClosed = true;
-            throw new NotImplementedException();
+            if (_isClosed)
+            {
+                return;
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:658]: Do we have a test that exercises this path?

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:658]: Not yet; we should.   I'm still tweaking some of the internals to flesh things out a bit more.   I'll add it before closing this out.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:674]: Added

**Pull request merged**